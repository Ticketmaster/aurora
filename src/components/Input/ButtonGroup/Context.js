import React from "react";

export const { Provider, Consumer } = React.createContext({
  selected: null,
  onChange: null
});

export const withButtonGroupConsumer = Component => props => (
  <Consumer>
    {({ onChange: onGroupChange, value: groupValue = {} }) => (
      <Component
        onGroupChange={onGroupChange}
        groupValue={groupValue}
        {...props}
      />
    )}
  </Consumer>
);
