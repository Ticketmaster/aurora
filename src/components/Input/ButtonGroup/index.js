/* eslint import/prefer-default-export: off */
export { default as ButtonGroup } from "./ButtonGroup";
