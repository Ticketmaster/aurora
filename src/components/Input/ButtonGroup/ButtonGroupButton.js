import React from "react";
import classnames from "classnames";
import styled from "styled-components";
import PropTypes from "prop-types";
import { Button } from "../../Button";
import composeEventHandlers from "../../../utils/composeEventHandlers";
import { Consumer } from "../../SelectionProvider/Context";
import { getThemeValue } from "../../../utils";
import spacing from "../../../theme/spacing";
import { typography } from "../../../theme";

const ButtonGroupButtonContainer = styled.div`
  width: 50%;
  &:nth-child(2n + 1) {
    padding-right: ${spacing.slim};
  }
  &:nth-child(2n) {
    padding-left: ${spacing.slim};
  }
`;

const StyledButton = styled(Button).attrs({
  variant: "outlineGray"
})`
  font-size: ${typography.size.kilo};
  font-weight: normal;
  margin-bottom: ${spacing.cozy};
  &:hover {
    border: solid 1px ${getThemeValue("brand")};
  }
  &:focus {
    box-shadow: none;
  }
  &.button__selected {
    background-color: ${getThemeValue("primary", "lightBase")};
    border: solid 1px ${getThemeValue("brand")};
  }
`;

const ButtonGroupButton = ({ children, value, onClick, ...props }) => (
  <Consumer>
    {values => (
      <ButtonGroupButtonContainer>
        <StyledButton
          onClick={composeEventHandlers(
            () => values.onClick({ value }),
            onClick
          )}
          {...props}
          className={classnames({
            button__selected: values.selected.includes(value)
          })}
        >
          {children}
        </StyledButton>
      </ButtonGroupButtonContainer>
    )}
  </Consumer>
);

ButtonGroupButton.defaultProps = {
  onClick: () => {}
};

ButtonGroupButton.propTypes = {
  value: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  children: PropTypes.node.isRequired
};

export default ButtonGroupButton;
