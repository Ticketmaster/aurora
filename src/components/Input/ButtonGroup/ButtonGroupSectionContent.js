import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import { withButtonGroupConsumer } from "./Context";
import SelectionProvider from "../../SelectionProvider/Provider";

const ButtonsSection = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

class ButtonGroupSectionContent extends PureComponent {
  static propTypes = {
    onGroupChange: PropTypes.func.isRequired,
    groupValue: PropTypes.objectOf(PropTypes.any).isRequired,
    sectionValue: PropTypes.string.isRequired,
    isMultiSelect: PropTypes.bool.isRequired,
    valueOverride: PropTypes.arrayOf(PropTypes.string),
    children: PropTypes.node.isRequired
  };

  static defaultProps = {
    valueOverride: null
  };

  onChange = (selected = []) => {
    const { onGroupChange, groupValue, sectionValue } = this.props;

    if (selected.length) {
      onGroupChange({ ...groupValue, [sectionValue]: selected });
      return;
    }

    onGroupChange({ ...groupValue, [sectionValue]: undefined });
  };

  render() {
    const {
      groupValue,
      sectionValue,
      isMultiSelect,
      valueOverride,
      children
    } = this.props;
    return (
      <ButtonsSection>
        <SelectionProvider
          onChange={this.onChange}
          isMultiSelect={isMultiSelect}
          value={groupValue[sectionValue]}
          valueOverride={valueOverride}
        >
          {children}
        </SelectionProvider>
      </ButtonsSection>
    );
  }
}

export default withButtonGroupConsumer(ButtonGroupSectionContent);
