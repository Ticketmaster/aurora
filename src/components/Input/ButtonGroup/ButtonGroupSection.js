import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import ButtonGroupSectionContent from "./ButtonGroupSectionContent";
import Spacing from "../../Spacing";
import { Text } from "../../Text";

const StyledText = styled(Text)`
  line-height: 1.25;
  text-transform: uppercase;
`;

const ButtonGroupSection = ({
  title,
  value: sectionValue,
  isMultiSelect,
  valueOverride,
  children
}) => (
  <Spacing bottom={{ xSmall: "moderate" }}>
    {title && (
      <Spacing bottom={{ xSmall: "moderate" }}>
        <StyledText secondary size="uno" weight="semiBold">
          {title}
        </StyledText>
      </Spacing>
    )}
    <ButtonGroupSectionContent
      isMultiSelect={isMultiSelect}
      sectionValue={sectionValue}
      valueOverride={valueOverride}
    >
      {children}
    </ButtonGroupSectionContent>
  </Spacing>
);

ButtonGroupSection.defaultProps = {
  title: "",
  isMultiSelect: false,
  valueOverride: null
};

ButtonGroupSection.propTypes = {
  title: PropTypes.string,
  value: PropTypes.string.isRequired,
  isMultiSelect: PropTypes.bool,
  valueOverride: PropTypes.arrayOf(PropTypes.string),
  children: PropTypes.node.isRequired
};

export default ButtonGroupSection;
