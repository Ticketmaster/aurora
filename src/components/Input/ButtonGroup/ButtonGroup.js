import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import ButtonGroupSection from "./ButtonGroupSection";
import ButtonGroupButton from "./ButtonGroupButton";
import { Provider as ButtonGroupProvider } from "./Context";
import spacing from "../../../theme/spacing";
import Spacing from "../../Spacing";
import { Text } from "../../Text";

const Group = styled.div`
  padding: ${spacing.normal} ${spacing.moderate} 0 ${spacing.moderate};
`;

const StyledText = styled(Text)`
  line-height: 1.25;
`;

const ButtonGroup = ({ title, children, value, onChange }) => (
  <Group>
    <Spacing bottom={{ xSmall: "moderate" }}>
      <StyledText size="giga" weight="semiBold">
        {title}
      </StyledText>
    </Spacing>
    <ButtonGroupProvider value={{ onChange, value }}>
      {children}
    </ButtonGroupProvider>
  </Group>
);

ButtonGroup.defaultProps = {
  value: {},
  onChange: () => {}
};

ButtonGroup.propTypes = {
  title: PropTypes.string.isRequired,
  value: PropTypes.shape({}),
  onChange: PropTypes.func,
  children: PropTypes.node.isRequired
};

ButtonGroup.Section = ButtonGroupSection;
ButtonGroup.Button = ButtonGroupButton;

export default ButtonGroup;
