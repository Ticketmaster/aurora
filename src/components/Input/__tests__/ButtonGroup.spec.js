import React from "react";
import { renderIntoDocument, cleanup, fireEvent } from "react-testing-library";
import ButtonGroup from "../ButtonGroup/ButtonGroup";

describe("ButtonGroup", () => {
  afterEach(cleanup);

  const renderButtonGroup = (groupProps, sectionProps) =>
    renderIntoDocument(
      <ButtonGroup title="Sort by" {...groupProps}>
        <ButtonGroup.Section title="Category" value="0" {...sectionProps}>
          <ButtonGroup.Button value="0" data-testid="test-button0">
            All
          </ButtonGroup.Button>
          <ButtonGroup.Button value="1" data-testid="test-button1">
            Concerts
          </ButtonGroup.Button>
          <ButtonGroup.Button value="2" data-testid="test-button2">
            Button Label
          </ButtonGroup.Button>
        </ButtonGroup.Section>
      </ButtonGroup>
    );
  it("renders default button group", () => {
    const component = renderButtonGroup();
    expect(component).toMatchSnapshot();
  });

  it("renders button group with default selected element", () => {
    const component = renderButtonGroup({ value: { 0: ["0"] } });
    expect(component).toMatchSnapshot();
    expect(component.getByTestId("test-button0").className).toEqual(
      expect.stringContaining("button__selected")
    );
  });

  it("selecting item", () => {
    const onChange = jest.fn();
    const component = renderButtonGroup({ onChange });
    fireEvent.click(component.getByTestId("test-button0"));
    expect(onChange).toHaveBeenCalledWith({ "0": ["0"] });
  });

  it("selecting item of button group with default selected item (no multi selection)", () => {
    const onChange = jest.fn();
    const component = renderButtonGroup({ value: { 0: ["0"] }, onChange });
    fireEvent.click(component.getByTestId("test-button1"));
    expect(onChange).toHaveBeenCalledWith({ "0": ["1"] });
  });

  it("selecting item of button group with several sections", () => {
    const onChange = jest.fn();
    const component = renderIntoDocument(
      <ButtonGroup title="Sort by" onChange={onChange} value={{ 0: ["0"] }}>
        <ButtonGroup.Section title="Category0" value="0">
          <ButtonGroup.Button value="0" data-testid="test-button0">
            All
          </ButtonGroup.Button>
        </ButtonGroup.Section>
        <ButtonGroup.Section title="Category1" value="1">
          <ButtonGroup.Button value="0" data-testid="test-button1">
            All
          </ButtonGroup.Button>
        </ButtonGroup.Section>
      </ButtonGroup>
    );
    fireEvent.click(component.getByTestId("test-button1"));
    expect(onChange).toHaveBeenCalledWith({ "0": ["0"], "1": ["0"] });
  });

  it("selecting multiple items of button group", () => {
    const onChange = jest.fn();
    const component = renderButtonGroup({ onChange }, { isMultiSelect: true });
    fireEvent.click(component.getByTestId("test-button0"));
    fireEvent.click(component.getByTestId("test-button1"));
    fireEvent.click(component.getByTestId("test-button2"));
    expect(onChange).toHaveBeenCalledWith({ "0": ["0", "1", "2"] });
    expect(onChange.mock.calls.length).toBe(3);
  });
});
