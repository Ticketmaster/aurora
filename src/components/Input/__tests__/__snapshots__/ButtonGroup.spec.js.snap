// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup renders button group with default selected element 1`] = `
Object {
  "container": <div>
    <div
      class="sc-kEYyzF cGSVRW"
    >
      <div
        class="sc-bwzfXH RpnLm"
      >
        <div
          class="text text--dark text--primary sc-kkGfuU kIGIsH sc-bZQynM dDBbro"
        >
          Sort by
        </div>
      </div>
      <div
        class="sc-bwzfXH RpnLm"
      >
        <div
          class="sc-bwzfXH RpnLm"
        >
          <div
            class="text text--dark text--secondary sc-jzJRlG kxFhHb sc-bZQynM dvqadL"
          >
            Category
          </div>
        </div>
        <div
          class="sc-bdVaJa erSRMf"
        >
          <div
            class="sc-eNQAEJ jfDpKy"
          >
            <button
              class="button__selected sc-hMqMXs jKUFUL noFocus sc-cSHVUG cpHQgj"
              data-testid="test-button0"
            >
              All
            </button>
          </div>
          <div
            class="sc-eNQAEJ jfDpKy"
          >
            <button
              class="sc-hMqMXs jKUFUL noFocus sc-cSHVUG cpHQgj"
              data-testid="test-button1"
            >
              Concerts
            </button>
          </div>
          <div
            class="sc-eNQAEJ jfDpKy"
          >
            <button
              class="sc-hMqMXs jKUFUL noFocus sc-cSHVUG cpHQgj"
              data-testid="test-button2"
            >
              Button Label
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ButtonGroup renders default button group 1`] = `
Object {
  "container": <div>
    <div
      class="sc-kEYyzF cGSVRW"
    >
      <div
        class="sc-bwzfXH RpnLm"
      >
        <div
          class="text text--dark text--primary sc-kkGfuU kIGIsH sc-bZQynM dDBbro"
        >
          Sort by
        </div>
      </div>
      <div
        class="sc-bwzfXH RpnLm"
      >
        <div
          class="sc-bwzfXH RpnLm"
        >
          <div
            class="text text--dark text--secondary sc-jzJRlG kxFhHb sc-bZQynM dvqadL"
          >
            Category
          </div>
        </div>
        <div
          class="sc-bdVaJa erSRMf"
        >
          <div
            class="sc-eNQAEJ jfDpKy"
          >
            <button
              class="sc-hMqMXs jKUFUL noFocus sc-cSHVUG cpHQgj"
              data-testid="test-button0"
            >
              All
            </button>
          </div>
          <div
            class="sc-eNQAEJ jfDpKy"
          >
            <button
              class="sc-hMqMXs jKUFUL noFocus sc-cSHVUG cpHQgj"
              data-testid="test-button1"
            >
              Concerts
            </button>
          </div>
          <div
            class="sc-eNQAEJ jfDpKy"
          >
            <button
              class="sc-hMqMXs jKUFUL noFocus sc-cSHVUG cpHQgj"
              data-testid="test-button2"
            >
              Button Label
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
